### DISCUSSION TITLE
X-Massge Header Problem

### DISCUSSION THREAD

**Discussion by Partymann2000** [fastapi/fastapi#11760] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
KMain Code
@app.post("/server/{guild_id}/modules/youtube/delete")
async def delete_youtube_channel(request: Request, 
                                 guild_id: int, 
                                 yt_channel_name: str = Form(...),
                                 dc_channel_id: int = Form(...),
                                 link: str = Form(...),
                                 message: str = Form(...)):
    session_id = request.cookies.get("session_id")
    if not session_id or not await db.get_session(session_id):
        raise HTTPException(status_code=401, detail="no auth")
    
    delete_youtube = await module_db.delete_youtube_channel(guild_id, yt_channel_name, dc_channel_id, link, message)

    print("----------------------------------")
    print(delete_youtube)
    print("----------------------------------")

    if delete_youtube is False:
        ...
    else:
        referer = request.headers.get("referer")
        headers = {"X-Message": "Channel erfolgreich gelöscht"}
        response = RedirectResponse(url=f"/server/{guild_id}/modules/youtube", headers=headers, status_code=303)
        print(response.headers)
        return response

It prints me the following: MutableHeaders({'x-message': 'Channel erfolgreich gelöscht', 'content-length': '0', 'location': '/server/852201283690692628/modules/youtube'})

this is my redering template
<p>{{ request.headers.get("x-message") }}</p>
but on the page the x header none is displayed as a p tag
```

### Description

I would like to add an X-Massge, but only None is displayed as the value

### Operating System

Windows

### FastAPI Version

latest

### Pydantic Version

latest

### Python Version

3.12

**Comment by chyok** [comment, marked as answer]
Hi @Partymann2000 ,
Controlling clients to maintain headers on the server side is not a feature supported by the HTTP protocol. This means that the server cannot require the browser to keep the original request's headers during the redirection process. When an HTTP redirect occurs, the browser initiates a new request, and unless the client (such as a browser or API client) actively maintains the headers, those headers will not be preserved.

To put it simply, the 303 response you received from your delete request does include the header you specified. However, the browser will not carry over this request header when redirecting to `modules/youtube`. That's why you cannot retrieve it using request.headers.get("x-message").

The alternative is to pass information using query parameters:
```python
...
    else:
	x_message = "Channel erfolgreich gelöscht"
        response = RedirectResponse(url=f"/server/{guild_id}/modules/youtube?x_message={x_message}", status_code=303)
        print(response.headers)
        return response


@app.get("/redirect-to-api")
async def redirect_to_target(x_message: str | None = None):
    return {"message": "You have been redirected.", "header": x_message}
```

**Comment by ceb10n** [comment]
If, **and only if** you can have some control over who is calling your api, then you can stop following redirects:

```python
@app.get("/redirect")
async def redirect():
    return {"hello": "redirect"}

@app.get("/")
async def will_redirect():
    x_message = "Channel erfolgreich gelöscht"
    headers = {"X-Message": x_message}
    return RedirectResponse(url="http://localhost:8000/redirect", headers=headers, status_code=303)
```

And then you disable redirects:

```python
import requests

r = requests.get("http://localhost:8000", allow_redirects=False)

print(f"Got my headers {r.headers}, now will follow redirect to {r.headers['Location']}")

r = requests.get(r.headers["Location"])

print(r.json())
```
